### PR TITLE
Add taker http basic auth

### LINKED ISSUE #609: Taker UI authentication
Implement basic auth for taker.

### PR DESCRIPTION
**Pull request by bonomat**
Resolves #609 

This is the first step to have authentication for the taker. 
For our current use of running the taker on umbrel this is sufficient. Eventually we might want to allow the user to change the password from inside of the application.

### CODE REVIEW

**Comment by bonomat** [comment]
Note: we should backport this to 0.3.x because it is mostly needed for https://github.com/getumbrel/umbrel/pull/1149

File: CHANGELOG.md [hunk 1]
```diff
@@ -16,6 +16,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
   The previous payout transactions are invalidated.
   The new payout transactions spend from the same lock transaction, so the rollover happens off-chain.
   In case a maker rejects a rollover request from a taker the old oracle price event and payout transactions stay in place.
+- Changed username for HTTP authentication to `itchysats`
```
Comment (thomaseizinger) [inline CHANGELOG.md:19]: This should go under a different section, we are in `### Added` here :)

File: daemon/src/auth.rs [hunk 2]
```diff
@@ -14,7 +14,8 @@ use std::str::FromStr;
 /// A request guard that can be included in handler definitions to enforce authentication.
 pub struct Authenticated {}
 
-pub const MAKER_USERNAME: &str = "maker";
+pub const USERNAME: &str = "itchysats";
+pub const TAKER_DEFAULT_PASSWORD: &str = "onbrinkofsecondbailout";
```
Comment (thomaseizinger) [inline daemon/src/auth.rs:18]: I'd rather not have a default password in the binary. If we want a default password for umbrel, we can always hardcode that in the docker-compose file. If `--password` is not specified, we can always derive a password from the seed.

What do you think?

File: daemon/src/routes_maker.rs [hunk 3]
```diff
@@ -304,12 +304,15 @@ mod tests {
     /// Creates an "Authorization" header that matches the password above,
     /// in particular it has been created through:
     /// ```
-    /// base64(maker:hex("Now I'm feelin' so fly like a G6"))
+    /// let password = "Now I'm feelin' so fly like a G6".to_string();
+    /// let password_vec = password.as_bytes();
+    /// let password_hex = hex::encode(vec);
+    /// let basic_auth = base64::encode(&format!("{}:{}", "itchysats", password_hex));
```
Comment (thomaseizinger) [inline daemon/src/routes_maker.rs:310]: That diff seems unnecessary if we are just changing the username from `maker` to `itchysats`.

File: daemon/src/routes_taker.rs [hunk 4]
```diff
@@ -259,3 +267,20 @@ pub async fn post_withdraw_request(
...
+    }
+}
+
+/// A rocket responder that prompts the user to sign in to access the API.
+#[derive(rocket::Responder)]
+#[response(status = 401)]
+pub struct PromptAuthentication {
+    inner: (),
+    www_authenticate: Header<'static>,
+}
```
Comment (thomaseizinger) [inline daemon/src/routes_taker.rs:286]: This could be re-used across both binaries.

File: daemon/src/routes_taker.rs [hunk 5]
```diff
@@ -259,3 +267,20 @@ pub async fn post_withdraw_request(
...
 }
+
+/// A "catcher" for all 401 responses, triggers the browser's basic auth implementation.
+#[rocket::catch(401)]
+pub fn unauthorized() -> PromptAuthentication {
+    PromptAuthentication {
+        inner: (),
+        www_authenticate: Header::new("WWW-Authenticate", r#"Basic charset="UTF-8"#),
+    }
+}
```
Comment (thomaseizinger) [inline daemon/src/routes_taker.rs:278]: This one probably too!

**Review by thomaseizinger — COMMENTED** [review]
Not really a fan of the default password IN the binary.

File: daemon/src/auth.rs (hunk 2 above)
Comment (klochowicz) [reply daemon/src/auth.rs:18]: also: if we would like a default password for the CI, we can always hardcode one in `cargo dev-taker` etc.

**Comment by thomaseizinger** [comment]
Closed in favor of https://github.com/itchysats/itchysats/pull/982.

**Pull request closed without merging**